[MRG] Improved the bluetooth scanning progress

huangyz0918 opened the pull request:
Closes #271 

<!-- 
Thank you for contributing to ODK!
-->

#### What has been done to verify that this works as intended?
I tested using Nexus 6P (android 8.0.1) everything went well.

The screenshots looks like:

|scanning page|sender page|
|:--:|:--:|
|![Screenshot_19700117-142625.png](https://i.loli.net/2019/07/21/5d33d66fc2a4256005.png)|![Screenshot_19700117-142830.png](https://i.loli.net/2019/07/21/5d33d66fd015584959.png)|

And I built a fixed time that bluetooth device is discoverable, in this way, we can know when we has finished the scanning and get the completed scanned results. An issue is should we notify the users that the discoverable time is 60s?

#### Why is this the best possible solution? Were any other approaches considered?
N/A
#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
N/A
#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkCode` and confirmed all checks still pass OR confirm CircleCI build passes
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/share/blob/master/share_app/src/main/assets/open_source_licenses.html).

lakshyagupta21 commented on skunkworks_crow/src/main/java/org/odk/share/bluetooth/BluetoothReceiver.java:
```diff
@@ -33,25 +33,27 @@ public void onReceive(Context context, Intent intent) {
-                    bluetoothReceiverListener.onDiscoveryFinished();
-                    break;
+        // if the discovery started, update the ui in activity.
+        if (BluetoothAdapter.ACTION_DISCOVERY_STARTED.equals(action)) {
```
I think `switch `case would be better in this case, it will make it easy to understand.

huangyz0918 replied:
I have an issue using `swith-case`, the `ACTION_DISCOVERY_STARTED` cannot be received, so I changed them to `if-else`, I think that is not a big deal.

huangyz0918 replied:
That's odd, maybe related to the `break`.

lakshyagupta21 replied:
That's strange, okay I'll take a look, anyways if you've added `if` statements then it would be better if you can make these individual stmts as `if else if..`, right now they all are written as individual `if` statements.

huangyz0918 replied:
Thanks, no hurry, let me have a try with `switch`.

lakshyagupta21 commented on skunkworks_crow/src/main/java/org/odk/share/views/ui/bluetooth/BtSenderActivity.java:
```diff
@@ -133,6 +136,16 @@ private Disposable addUploadEventSubscription() {
+    /**
+     * Enable the bluetooth discovery for other devices. The timeout is 60 seconds.
+     */
+    private void enableDiscovery() {
```
And I think sender should be made `DISCOVERABLE` not the receiver so that it would be able to see the sender in its scanned results.

huangyz0918 replied:
This method allows sender to be `DISCOVERABLE` for a specific duration, by default, it is 120s. And I set a fixed 60s for that.  I think this change can help receiver to show all the possible senders and help them establish a valid connection in a fixed time, since in the next step we are going  to build a default method, we need that timeout I think. 

As for who should scan bluetooth devices, I think sender should host a file server so it should be `DISCOVERABLE`, the connect invitation should be sent by the receiver.

lakshyagupta21 replied:
I don't think just by hosting the file the sender is made discoverable.  Please check that on multiple devices.

huangyz0918 replied:
Bluetooth doesn't support multiple devices, the files server should be ready first, so the sender should get ready before the receiver. The connection order should start with the sender then the receiver, so letting receiver to connect a ready server is no issues.

lakshyagupta21 replied:
By multiple devices, I mean to say that try to test this behavior in different devices, not by running them all at once.

lakshyagupta21 commented on skunkworks_crow/src/main/res/values/strings.xml:
```diff
@@ -139,8 +139,10 @@
-    <string name="bluetooth_unpaired">unpaired, reachable</string>
+    <string name="scanning_title">Scanning</string>
+    <string name="scanning_msg">Scanning bluetooth devices, please wait&#8230;</string>
+    <string name="bluetooth_paired">paired</string>
```
It could be like `Device Paired` or `Device Unpaired`.

huangyz0918 replied:
Sure, will fix right now.

huangyz0918 commented:
@lakshyagupta21 @shobhitagarwal1612 please review, thanks.

huangyz0918 commented on skunkworks_crow/src/main/java/org/odk/share/bluetooth/BluetoothReceiver.java:
```diff
@@ -33,24 +33,26 @@ public void onReceive(Context context, Intent intent) {
+        switch (action) {
+            case BluetoothAdapter.ACTION_DISCOVERY_STARTED:
+                // if the discovery started, update the ui in activity.
+                Timber.d("bluetooth devices discovery started...");
```
Hmm.. It seems `ACTION_DISCOVERY_STARTED` should write before `ACTION_FOUND`, I'm not sure, but it works now @lakshyagupta21 .

huangyz0918 commented:
New updates:

- Added a count down timer for discoverable time.
- Added a cancel button while connecting to bluetooth device.

@lakshyagupta21 @shobhitagarwal1612 This is an important PR, and I think it will be good if you can merge ASAP, thanks.
